src/Misc/tools.ts

```diff
@@ -1215,6 +1215,14 @@ export class AsyncLoop {
             }
         }, callback);
     }
+
+    /**
+     * Utility function to detect if the current user agent is Safari
+     * @returns whether or not the current user agent is safari
+     */
+    public static IsSafari() : boolean {
+        return /^((?!chrome|android).)*safari/i.test(navigator.userAgent);

```

Comment:
kudos of some logic in the thingine that is used to detect certain bad desktop OSes. 😉

I opted to add this logic to Tools as well, but perhaps I should modify thinEngine to reuse the logic there instead?

Reply:
We need to keep thinengine with no dependency so let's keep it like that for now